src/python/foglamp/core/api/browser.py

```diff
@@ -70,77 +63,64 @@ def setup(app):
-        results.append(dict(zip(columns, row)))
+    agg = {"operation": "count", "column": "*", "alias": "Count"}
+    d = OrderedDict()
+    d['aggregate'] = agg
+    d['group'] = "asset_code"
 
-    # Close the connection.
-    await conn.close()
+    _storage = connect.get_storage()
+    results = _storage.query_tbl_with_payload('readings', json.dumps(d))
 
-    return web.json_response(results)
+    return web.json_response(results['rows'])
 
 
 async def asset(request):
-    """
-    Browse a particular asset for which we have recorded readings and
+    """ Browse a particular asset for which we have recorded readings and
     return a readings with timestamps for the asset. The number of readings
     return is defaulted to a small number (20), this may be changed by supplying
     the query parameter ?limit=xx&skip=xx
 
-    Return the result of the Postgres query 
+    Return the result of the Postgres query
     SELECT TO_CHAR(user_ts, '__TIMESTAMP_FMT') as "timestamp", (reading)::jsonFROM readings WHERE asset_code = 'asset_code' ORDER BY user_ts DESC LIMIT 20 OFFSET 0
     """
-
-    conn = await asyncpg.connect(**__CONNECTION)
     asset_code = request.match_info.get('asset_code', '')
 
-    query = """
-            SELECT TO_CHAR(user_ts, '{0}') as "timestamp", (reading)::json
-            FROM readings WHERE asset_code = '{1}'
-            """.format(__TIMESTAMP_FMT, asset_code)
-
-    query += _where_clause(request)
+    # TODO: FOGL-637, 640
+    timestamp = {"column": "user_ts", "format": __TIMESTAMP_FMT, "alias": "timestamp"}
+    d = OrderedDict()
+    d['return'] = [timestamp, "reading"]
+    d['where'] = {"column": "asset_code", "condition": "=", "value": asset_code}
+    _and = _where_clause(request)
+    if len(_and) > 0:
+        d['where'].update(_and)
 
     # Add the order by and limit clause
-    limit = __DEFAULT_LIMIT
-    offset = __DEFAULT_OFFSET
-    if 'limit' in request.query:
-        limit = request.query['limit']
-        offset = request.query['skip'] if 'skip' in request.query else __DEFAULT_OFFSET
+    d['sort'] = {"column": "user_ts", "direction": "desc"}
+    limit = int(request.query.get('limit')) if 'limit' in request.query else __DEFAULT_LIMIT
+    offset = int(request.query.get('skip')) if 'skip' in request.query else __DEFAULT_OFFSET
 
-    orderby = ' ORDER BY user_ts DESC LIMIT {0} OFFSET {1}'.format(limit, offset)
-    query += orderby
+    d['limit'] = limit
+    if offset:
+        d['skip'] = offset
 
-    # Select the assets from the readings table
-    rows = await conn.fetch(query)
-    results = []
-    for row in rows:
-        jrow = {'timestamp': row['timestamp'], 'reading': json.loads(row['reading'])}
-        results.append(jrow)
+    _storage = connect.get_storage()
+    results = _storage.query_tbl_with_payload('readings', json.dumps(d))
 
-    # Close the connection.
-    await conn.close()
-
-    return web.json_response(results)
+    return web.json_response(results['rows'])
```

Comment:
in case of error : return web.json_response(results['message']) Here HTTP status code is 200 itself

Reply:
We will handle this as separately task. for all end-points

Reply:
Please add task/TODO for this. Then we can approve and get this merged in.